<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Не-зашифрованная" data-toc-modified-id="Не-зашифрованная-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Не зашифрованная</a></span></li><li><span><a href="#Зашифрованная" data-toc-modified-id="Зашифрованная-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Зашифрованная</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Вывод</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

**Заказчик** — страховая компания «Хоть потоп».

**Цель исследования** 

* Поиск наиболее оптимального медота шифрования данных пользователей


**Входные данные**  — ```/datasets/insurance.csv``` - база данных пользователей


**Результаты исследования** — 

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
# выгрузим данные и посмотрим в каком виде получается датафрейм
df = pd.read_csv('/datasets/insurance.csv')
df.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

**Признаки:** пол, возраст и зарплата застрахованного, количество членов его семьи.

**Целевой признак:** количество страховых выплат клиенту за последние 5 лет.

In [3]:
# изучим датафрейм
# функция для обработки данных датафрейма
def research(df):
    
    display(df)
    print('Информация по таблице:', '\n')
    df.info()
    print()
    print('Количество пропусков:', '\n')
    print(df.isna().sum(), '\n', '\n')
    print('Процент пропусков:', '\n')
    print(df.isna().mean()*100, '\n')

    display(df.describe())
    print()
    print('Число явных дубликатов:', df.duplicated().sum(), '\n')


In [4]:
research(df)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0       1     41.0   49600.0            1                  0
1       0     46.0   38000.0            1                  1
2       0     29.0   21000.0            0                  0
3       0     21.0   41700.0            2                  0
4       1     28.0   26100.0            0                  0
...   ...      ...       ...          ...                ...
4995    0     28.0   35700.0            2                  0
4996    0     34.0   52400.0            1                  0
4997    0     20.0   33900.0            2                  0
4998    1     22.0   32700.0            3                  0
4999    1     28.0   40600.0            1                  0

[5000 rows x 5 columns]

Информация по таблице: 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB

Количество пропусков: 

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64 
 

Процент пропусков: 

Пол                  0.0
Возраст              0.0
Зарплата             0.0
Члены семьи          0.0
Страховые выплаты    0.0
dtype: float64 



Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000


Число явных дубликатов: 153 



In [5]:
# Посмотрим как распределились данные
df["Пол"].value_counts()

0    2505
1    2495
Name: Пол, dtype: int64

**Промежуточный итог:**

* Пропусков нет
* в идеале поменять названия столбцов на змеиный регистр или верблюжий.
* имеются 153 явных дубликатов (но так как в данной таблице нет уникального идентификатора, представленные данные скорее всего не дубликаты)
* распределение почти равное, перевеса нет
* минимальная зарплата в 5300 - грустно


В остальном данные адекватные.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Признаки умножают на обратимую матрицу. Изменится ли качество линейной регрессии? (Её можно обучить заново.)**

**Ответ:** Не изменится

**Обоснование:** ...

Признаки умножают на обратимую матрицу $P$ :

$$
w_1 = ((XP)^T XP)^{-1} (XP)^T y
$$

Раскроем скобки

$$
w_1 = (P^TX^T XP)^{-1} P^TX^T y
$$

Используем следующую формулу
☑

$$
(A*B)^{-1} =  B^{-1} * A^{-1} 
$$
где: $A$ и $B$ квадратные

$P$, $P^T$ и $X$, $X^T$ - квадратные $=>$

$$
w_1 = P^{-1}(X^T X)^{-1} (P^T)^{-1}  P^TX^T y
$$

Умножение на обратную матрицу, равно еденичной матрице

$$
w_1 = P^{-1} (X^T X)^{-1} IX^T y
$$

Умножение матрицы на единичную, равно этой матрицы

$$
w_1 = P^{-1} (X^T X)^{-1} X^T y
$$

$$
a_1=XPw_1
$$

$$
a_1 = XP P^{-1} (X^T X)^{-1} X^T y
$$

$$
a_1 = X (X^T X)^{-1} X^T y
$$

$$
a_1 = Xw
$$



## Алгоритм преобразования

Предложите алгоритм преобразования данных для решения задачи. Обоснуйте, почему качество линейной регрессии не поменяется.

**Алгоритм**

Для кодирования нашей матрицы, умножим на **обратимую*** матрицу (4х4) и в результате должны получить новую матрицу (5000х4)

**Обоснование:**

In [6]:
# разобьем df на признаки и целевые признаки согласно заданию
features = df.drop('Страховые выплаты', axis=1)
target = df['Страховые выплаты']

In [7]:
# Размеры обратимой* матрицы завист от столбцов features и заполним их рандомными значениями
state = np.random.RandomState(12345)
inverse_matrix = state.normal(size=(features.shape[1],features.shape[1]))
inverse_matrix

array([[-0.20470766,  0.47894334, -0.51943872, -0.5557303 ],
       [ 1.96578057,  1.39340583,  0.09290788,  0.28174615],
       [ 0.76902257,  1.24643474,  1.00718936, -1.29622111],
       [ 0.27499163,  0.22891288,  1.35291684,  0.88642934]])

try:
    np.linalg.inv(inverse_matrix)
except:
    display('Матрица необратимая')

In [11]:
rs = np.random.RandomState(42)
def matrix_generator(features, rs):
    m = rs.normal(size=(features.shape[1],features.shape[1]))
    try:
       np.linalg.inv(m)
       return m
    except:
        return matrix_generator(features, rs)

In [8]:
encrypted_features = features @ inverse_matrix
encrypted_features

0             1             2             3
0     38224.186641  61881.000420  49961.234837 -64280.684721
1     29313.558467  47428.845564  38278.822267 -49242.555394
2     16206.481556  26215.538233  21153.670838 -27212.472653
3     32110.072445  52006.047856  42004.453110 -54044.730722
4     20126.326163  32571.440926  26289.724215 -33824.037786
...            ...           ...           ...           ...
4995  27509.697503  44537.193278  35961.967320 -46265.431844
4996  40363.894074  65360.784898  52781.234122 -67911.520319
4997  26109.730637  42282.463506  34148.283213 -43934.487817
4998  25190.905401  40790.236490  32940.675278 -42378.128295
4999  31277.428385  50644.973517  40895.322817 -52618.357439

[5000 rows x 4 columns]

Стоит помнить что в питоне есть несколько способов получить произведение матриц (`numpy.dot()`, `numpy.multiply()`, `numpy.matmul()`, через операторы  `*` и `@`), но нужно учитывать, что результаты некоторых из них могут отличаться.

Выбирать нужно в зависимости от задач:
* `numpy.multiply()` - это поэлементное перемножение матриц (новая матрица заполняется произведениями элементов с одинаковыми индексами), аналог оператора `*`

* `numpy.matmul()` - аналог оператора `@`

* `numpy.dot()` - на двумерных матрицах выдает тот же результат, что и `numpy.matmul()` и `@`, но на трехмерных и более многомерных матрицах результаты уже будут отличаться


**Прмоежуточный вывод:**

Количество столбцов и строк не изменилось (размер матрицы совпадает), можем проверять модели

## Проверка алгоритма

### Не зашифрованная

In [9]:
model = LinearRegression()

model.fit(features, target)

predictions = model.predict(features)

print('R2:', r2_score(target,predictions))

R2: 0.4249455028666801


### Зашифрованная

In [10]:
model = LinearRegression()

model.fit(encrypted_features, target)

predictions = model.predict(encrypted_features)

print('R2:', r2_score(target,predictions))

R2: 0.42494550286667687


## Вывод

Качество модели практически не изменилось, есть небольшое отличие в e-16 (не совсем понимаю из-за чего может быть, возможно связано с размерностью зашифрованной таблицы, так как в оригинальной версии больше целых значений)

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования